### ISSUE TITLE
[ARM64] Incorrect HFA/HVA property calculation

### ISSUE THREAD

**Issue by AntonLapounov** [dotnet/runtime#35144]
While reviewing changes in dotnet/coreclr#23675, I noticed that the code added to `EEClass::CheckForHFA` does not handle wrapped `Vector64`, `Vector128`, and  `Vector256` intrinsic types correctly.  For instance, it does not distinguish a wrapped `Vector64` and a `double`.  Moreover, the `elemSize` check is skipped for wrapped `Vector128` and  `Vector256` types.  As a result, the HFA/HVA property may be calculated incorrectly.  For instance (here `{ }` denotes a struct):
* `{{Vector64}, Vector64}` is incorrectly treated as non-HVA.
* `{{Vector64}, double}` is incorrectly treated as HFA(double).
* `{Vector128, {Vector256}}` is incorrectly treated as HVA(simd16).

The code in question: https://github.com/dotnet/runtime/blob/ce416f4c8ba1556bbb8e3d0b0a1a3ee7f16fbc7f/src/coreclr/src/vm/class.cpp#L1345-L1381

The repro program (`set COMPlus_JITDisasm=C::*` to see HFA/HVA properties and registers used):
```C#
using System.Runtime.CompilerServices;
using System.Runtime.Intrinsics;

#pragma warning disable 0169 // warning CS0169: The field '{0}' is never used
struct WrappedVector64  { Vector64<byte> _; }
struct WrappedVector128 { Vector128<byte> _; }
struct WrappedVector256 { Vector256<byte> _; }

// Incorrectly treated as non-HVA: passed in x0, x1
struct S1 { WrappedVector64 x; Vector64<byte> y; }

// Incorrectly treated as HFA(double): passed in d0, d1
struct S2 { WrappedVector64 x; double y; }

// Incorrectly treated as HVA(simd16): passed in q0, q1, q2
struct S3 { Vector128<byte> x; WrappedVector256 y; }

static class C
{
    [MethodImpl(MethodImplOptions.NoInlining)]
    static void Foo<T>(T x) { }

    static void Main()
    {
        Foo(new S1());
        Foo(new S2());
        Foo(new S3());
    }
}
```
@CarolEidt @echesakovMSFT @sdmaclea

**Comment by sdmaclea** [comment]
@AntonLapounov and I spent a lot of time discussing this earlier this week ago.  I believe his assertions above are correct.

It should also be noted that @AntonLapounov compared VC++ compiler and our JIT regarding handling doubles and vector types to confirm our analysis.

**Comment by davidwrighton** [comment]
@CarolEidt needs to be aware of this.

**Comment by sdmaclea** [comment]
For the `Vector256<T>` case, it is a bit unclear.  Since it is not a hardware supported SIMD type, our handling is debatable.  `Vector256<T>` could be treated as if it were `{Vector128<T>, Vector128<T>}`

**Comment by CarolEidt** [comment]
Thanks @davidwrighton - I saw this. This is unfortunate as it will require a more complex interaction between the VM and the JIT. I suspect it needs some VM/JIT collaboration to address.

**Comment by AntonLapounov** [comment]
For reference, the C++ definitions below are equivalent to the `S1` and the `S2` managed cases above and demonstrate correct behavior of VC++ and GCC compilers regarding the [ABI spec](https://github.com/ARM-software/abi-aa/releases).  There is no equivalent of the `S3` case.
```C++
#include <arm_neon.h>

// HVA(__n64): passed in d0, d1 registers
struct S1 { 
    struct { uint8x8_t z; } x;
    uint8x8_t y;
};

// Non-HFA/HVA: passed in x0, x1 registers
struct S2 { 
    struct { uint8x8_t z; } x;
    double y;
};

void Foo(S1 x) { }
void Foo(S2 x) { }
```

**Comment by AntonLapounov** [comment]
> `Vector256<T>` could be treated as if it were an HVA `{Vector128<T>, Vector128<T>}`.

Yes, but that might be not future-proof if one day Neon registers are extended to 256 bits.

**Comment by sdmaclea** [comment]
>  be not future-proof

Possibly true.

Although we have Arm64 feature bits to controll what architecture we are targeting. 

> if one day Neon registers are extended to 256 bits.

It is unlikely.  The next generation SIMD for ARM is the scalable vector extension. It generalizes the handling of vectors. It is designed to allow hardware to add support for longer vectors (up to 2048 bits).  

The extending neon registers would require new instructions. As neon encodes the vector length in the instruction.

**Comment by AntonLapounov** [comment]
@sdmaclea Sounds reasonable. Anyway, we still have a bug that the two structs below are handled differently at present.  That is caused by ignoring the size of the nested `Vector256<byte>` field for `S3`, but taking it into account for `S4`.
```C#
// Treated as HVA(simd16): passed by value in q0, q1, q2
struct S3 { Vector128<byte> x; WrappedVector256 y; }

// Treated as non-HVA: passed by reference in x0
struct S4 { Vector128<byte> x; Vector256<byte> y; }
```

**Comment by davidwrighton** [comment]
Calling conventions for Vector256 is quite complex. For instance on X64, the appropriate calling convention changes based on whether or not the architecture has AVX support. I'm slowing working on a fix to make the runtime/JIT handle this correctly, but its slow going.
